### PR TITLE
Release 0.2.0

### LINKED ISSUE #23: Typo in README
README says 

```
from bedms.constants import AVAILABLE_SCHEMAS
print(AVAILABLE_SCHEMAS)

# >> ['ENCODE', 'FAIRTRACKS', 'BEDBASE'] 

```

What actually works is

```
from bedms.const import AVAILABLE_SCHEMAS
print(AVAILABLE_SCHEMAS)

# >> ['ENCODE', 'FAIRTRACKS', 'BEDBASE'] 

```

### PR DESCRIPTION
**Pull request by saanikat**
## Changes:
- Fixed #23 
- Added generic way of initialization of all schemas
- Added `TrainStandardizer` module for custom model creation

---
- [x] Update version
- [x] Chnagelog

### CODE REVIEW

**Comment by saanikat** [comment]
Added `TrainStandardizer` for training custom models by the user. 
The present `README.md` doesn't have the details, documentation will be added to [bedbase docs](https://docs.bedbase.org/bedboss/)

**Comment by saanikat** [comment]
### Separation of schemas from BEDMS

Available schemas for standardization have been moved to the HuggingFace repository :https://huggingface.co/databio/attribute-standardizer-model6
This solves the issue of having to update BEDMS each time a new schema is added by us. 
`README.md` has been updated with the new function calls. 
Earlier we would instantiate it like like:
```python
from bedms import AttrStandardizer

model = AttrStandardizer("ENCODE")
```
BEDMS had mapped schema name `ENCODE` to the model, and its associated configuration and files. Similarly, `BEDBASE` and `FAIRTRACKS` were associated with their respective files. However, this would've required us to update the package each time we added a new schema. 
Now, we need to provide the schema model and its associated configuration to BEDMS via HuggingFace. In the HuggingFace repository, each schema has its own directory. And each time a schema is added, a new schema directory would be added to HuggingFace ( details of adding a new schema have been provided there). The instantiation looks like this:
```python
from bedms import AttrStandardizer

model = AttrStandardizer(
    repo_id="databio/attribute-standardizer-model6", model_name="encode"
)
```
This also makes it easier for the user to provide their chosen schemas ( as long as they have models on their HuggingFace repository ).

**Comment by sanghoonio** [comment]
Is it worth updating 

`from bedms.const import AVAILABLE_SCHEMAS` 

to return a dictionary that includes the repo_id value for the 3 schemas we provide?

Or should we just hardcode the 3 repo IDs from PEPhub?

File: README.md [hunk 1]
```diff
@@ -16,22 +15,69 @@ pip install git+https://github.com/databio/bedms.git
...
+To load the datasets and encode them:
+
+```python
+trainer.load_encode_data()
+```
+
+To train the custom model:
+
+```python
+trainer.training()
```
Comment (nleroy917) [inline README.md]: Should this be `.train()`?

File: README.md [hunk 2]
```diff
@@ -16,22 +15,69 @@ pip install git+https://github.com/databio/bedms.git
...
+```python
+from bedms.train import AttrStandardizerTrainer
+
+trainer = AttrStandardizerTrainer("training_config.yaml")
+
+```
+To load the datasets and encode them:
+
+```python
+trainer.load_encode_data()
```
Comment (nleroy917) [inline README.md]: Does this return anything? I'd expect this to return something

File: README.md [hunk 3]
```diff
@@ -16,22 +15,69 @@ pip install git+https://github.com/databio/bedms.git
...
+
+```python
+trainer.training()
+```
+
+To test the custom model:
 
-To see the available schemas, you can run:
+```python
+trainer.testing()
```
Comment (nleroy917) [inline README.md]: Again here I like `.test()` over `.testing()`

File: README.md [hunk 4]
```diff
@@ -16,22 +15,69 @@ pip install git+https://github.com/databio/bedms.git
...
+
+```python
+from bedms import AttrStandardizer
+
+model = AttrStandardizer(
+    repo_id="name/of/your/hf/repo", model_name="model/name"
+)
+results = model.standardize(pep="geo/gse228634:default")
+
+assert results
```
Comment (nleroy917) [inline README.md]: Instead of `assert` could we show how maybe they could iterate over results? Then maybe include comments on what the expected output of a `print` might be? Something like:
```python
for result in results:
    print(result) # {'attr': 'genome', 'score': 0.8 }
```

File: bedms/train.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        self.train_accuracies = None
+        self.val_accuracies = None
+        self.train_losses = None
+        self.val_losses = None
+        self.model = None
+        self.fpr = None
+        self.tpr = None
+        self.roc_auc = None
+        self.all_labels = None
+        self.all_preds = None
```
Comment (nleroy917) [inline bedms/train.py]: If possible, could we type these?

File: bedms/train.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        logger.info("Encoding Done.")
+
+        batch_size = self.config["training"]["batch_size"]
+        self.train_loader = data_loader(train_encoded_data, batch_size)
+        self.test_loader = data_loader(test_encoded_data, batch_size)
+        self.val_loader = data_loader(val_encoded_data, batch_size)
+
+        logger.info("Loading Done.")
+
+    def training(self):
```
Comment (nleroy917) [inline bedms/train.py]: As mentioned in the README, I prefer `.train()` here. Feels more like a call to action, where `.training()` almost feels like i am checking or setting the status of something

File: bedms/train.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            device,
+            num_epochs,
+            self.output_size,
+            model_pth,
+            bow_drops,
+        )
+
+        logger.info("Training Done.")
+
+    def testing(self):
```
Comment (nleroy917) [inline bedms/train.py]: Same here, I'd prefer `.test()`

File: bedms/train.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        """
+        Model testing.
+        """
+        self.all_preds, self.all_labels = model_testing(
+            self.model, self.test_loader, self.criterion
+        )
+        precision = precision_score(self.all_labels, self.all_preds, average="macro")
+        recall = recall_score(self.all_labels, self.all_preds, average="macro")
+        f1 = f1_score(self.all_labels, self.all_preds, average="macro")
+        logger.info(f"Precision:{precision}, Recall: {recall}, F1 Score: {f1}")
```
Comment (nleroy917) [inline bedms/train.py:258]: I would return these values. the user probably wants them for themselves anyways

Comment (nleroy917) [reply bedms/train.py:258]: Like:
```python
return {
  'precision': precision,
  'recall': recall,
  'f1': f1
}
```

File: bedms/train.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+)
+from .const import PROJECT_NAME
+from .model import BoWSTModel
+
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(PROJECT_NAME)
+
+
+class AttrStandardizerTrainer:
```
Comment (nleroy917) [inline bedms/train.py:37]: I like that class, but the fact that no methods take any arguments makes me think they should just run automatically. The user has no control over anything so they now need to just run the functions... For example:

```python
trainer = AttrStandardizerTrainer("path/to/config.yaml")
trainer.load_encode_data()
trainer.train()
trainer.test()
```
I suppose the `train` and `test` **should not** run automatically of course, but I guess I'll argue that i think `load_encode_data` should just auto-run since they have to do it anyways... unless its a ton of data

File: bedms/train.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+            )
+            return
+
+        total_files = len(values_files_list)
+
+        paired_files = list(zip(values_files_list, headers_files_list))
+
+        train_size = self.config["data_split"]["train_set"]
+        test_size = self.config["data_split"]["test_set"]
+        val_size = self.config["data_split"]["val_set"]
```
Comment (nleroy917) [inline bedms/train.py:113]: I think that these should be percentages, not counts of files.. like i want to specify `0.7`, `0.2`, and `0.1`

File: bedms/train.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+                f"train({train_size}) + val({val_size}) + \
+                test({test_size}) > total_files({total_files})"
+            )
+            return
+
+        train_files = paired_files[:train_size]
+        val_files = paired_files[train_size : train_size + val_size]
+        test_files = paired_files[
+            train_size + val_size : train_size + val_size + test_size
+        ]
```
Comment (nleroy917) [inline bedms/train.py:127]: Then down here you can say:
```python
num_train_files = int(train_size * len(paired_files))
train_files = paired_files[:num_train_files]
```

Also now looking at this I'd make sure to shuffle `paired_files` before calling this code; that way you aren't getting big class imbalances in your training and test/validation sets

File: bedms/train.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        self.val_losses = None
+        self.model = None
+        self.fpr = None
+        self.tpr = None
+        self.roc_auc = None
+        self.all_labels = None
+        self.all_preds = None
+
+        with open(config, "r") as file:
+            self.config = yaml.safe_load(file)
```
Comment (nleroy917) [inline bedms/train.py:68]: Can we validate the config after this? there are a lot of functions that dig into it without really checking all the values are there, so maybe a short `_validate_config` to make sure the config the user provided is good and wont cause issues so they can get immediate feedback

File: bedms/utils_train.py [hunk 13]
```diff
@@ -0,0 +1,692 @@
...
+    train_losses = []
+    val_losses = []
+
+    best_val_loss = float("inf")
+    best_epoch = 0
+    early_stop = False
+
+    model.train()
+
+    for epoch in range(num_epochs):
```
Comment (nleroy917) [inline bedms/utils_train.py:444]: Can we wrap this in a `tqdm` so users can track the progress?

File: bedms/utils_train.py [hunk 14]
```diff
@@ -0,0 +1,692 @@
...
+        total_samples = 0
+        correct_predictions = 0
+        train_loss = 0.0
+        for x_values_bow, x_values_embeddings, x_headers_embeddings, y in train_loader:
```
Comment (nleroy917) [inline bedms/utils_train.py:448]: `tqdm` here too?

File: bedms/utils_train.py [hunk 15]
```diff
@@ -0,0 +1,692 @@
...
+    """
+    Loads each file from the directory path.
+
+    :param str dir: Path to the directory.
+    :return: List:paths to each file in the directory.
+    """
+    return glob(os.path.join(dir, "*.csv"))
+
+
+def load_and_preprocess(file_path: str) -> pd.DataFrame:
```
Comment (nleroy917) [inline bedms/utils_train.py]: More descriptive name? What are we loading and preprocessing?

File: bedms/utils_train.py [hunk 16]
```diff
@@ -0,0 +1,692 @@
...
+logger = logging.getLogger(PROJECT_NAME)
+
+warnings.filterwarnings(
+    "ignore",
+    category=UserWarning,
+    message="Creating a tensor from a list of numpy.ndarrays is extremely slow.",
+)
+
+
+def load_from_dir(dir: str) -> List[str]:
```
Comment (nleroy917) [inline bedms/utils_train.py]: Same here can this be a bit more descriptive?

File: bedms/utils_train.py [hunk 17]
```diff
@@ -0,0 +1,692 @@
...
+) -> torch.Tensor:
+    """
+    Computes the clutser-averaged top training embeddings using k-means clustering.
+
+    :param List[torch.tensor] embeddings: List of embedding tensors to cluster.
+    :param int num: Number of clusters to be created using k-means.
+    :return torch.Tensor: A tensor representing the
+        average of embeddings in the most common cluster.
+    """
+    flattened_embeddings = [embedding.tolist() for embedding in embeddings]
```
Comment (nleroy917) [inline bedms/utils_train.py]: I dont think this is flattened, its just converting the list of `np.array`s to a list of lists, right?

File: bedms/utils_train.py [hunk 18]
```diff
@@ -0,0 +1,692 @@
...
+    ) = encoded_data
+    # Convert data to TensorDataset
+    dataset = TensorDataset(
+        x_values_bow_tensor,
+        x_values_embeddings_tensor,
+        x_headers_embeddings_tensor,
+        y_tensor,
+    )
+    # Create DataLoader
+    return DataLoader(dataset, batch_size=batch_size, shuffle=True)
```
Comment (nleroy917) [inline bedms/utils_train.py:362]: Specify num_workers?

File: bedms/utils_train.py [hunk 19]
```diff
@@ -0,0 +1,692 @@
...
+            all_preds.extend(predicted_test.cpu().numpy())
+            all_labels.extend(labels.cpu().numpy())
+    test_accuracy = total_correct_test / total_samples_test
+    test_loss = total_loss_test / len(test_loader)
+    logger.info(f"Test Accuracy: {test_accuracy}, Test Loss: {test_loss}")
+
+    return all_preds, all_labels
+
+
+def plot_learning_curve(
```
Comment (nleroy917) [inline bedms/utils_train.py:583]: What do we think about returning the plots to the user?

**Review by nleroy917 — CHANGES_REQUESTED** [review]
I think this is good for the most part. Small comments everywhere. There was an overarching theme I just wanted to bring up... curious what others think:

### Using classes as state-storage
I brought this up in a recent review for @ClaudeHu. I'm noticing a pattern where people define and call functions on classes that set state _inside_ that class. So for example in the `AttributeStandardizerTrainer`, we call `load_enode_data`...

This doesn't return anything, instead if sets state on the trainer itself. In essence, it does this:
```python
def load_encode_data(self):
    self.encode_data = ...
```
Instead of:
```python
def load_encode_data(self):
    encode_data = ...
    return encode_data
```
I don't know if one is better than the other, but the former feels like an anti-pattern to me for some reason. I suppose my argument for the second version is that it gives control of the data to the user and reduces ambiguity about whats going on; it hides less functionality.

### Last comments
Lastly, is it worth considering `lightning`? I know the gains are smaller when there's no deep learning or multi-GPU training going on, but it really does clean up a lot of stuff and handle boring nuances that don't matter. Its a nice paradigm.

Moreover, if possible, we could integrate ClearML logging so the user can view the training in real-time instead of waiting for the plot to be saved to disk only to realize their model didn't learn anything...

Again this is if training times are long, if not then we can forget it.

File: bedms/attr_standardizer.py [hunk 20]
```diff
@@ -108,63 +92,54 @@ def _load_model(self) -> Tuple[nn.Module, object, object]:
...
-            vectorizer = None
-            label_encoder = None
-
-            vc_path = hf_hub_download(
-                repo_id=REPO_ID,
-                filename=filename_vc,
-            )
+        model_filename = f"model_{self.model_name}.pth"
+        label_encoder_filename = f"label_encoder_{self.model_name}.pkl"
+        vectorizer_filename = f"vectorizer_{self.model_name}.pkl"
```
Comment (khoroshevskyi) [inline bedms/attr_standardizer.py:97]: I like it! Very nice generic solution

**Review by khoroshevskyi — COMMENTED** [review]
All changes are really good, and I like that code is not duplicated and simple.

One thing before merging, can we clean .git history/folder, because it is 100mb?
